### PR TITLE
Raster calculator crash and hang fixes. Fixes #20583 #19760

### PR DESCRIPTION
**Pull request by elpaso**
This fixes a crash and a GUI issue, the second one only partially because the feedback while reading the raster block was disabled and I don't know why: https://github.com/qgis/QGIS/blob/master/src/providers/gdal/qgsgdalproviderbase.cpp#L282 @nyalldawson any hope to get this cancellation back?

### CODE REVIEW

**Comment by nyalldawson** [comment]
LGTM

Regarding the feedback issue - @wonder-sk or @rouault would be better to ask, I'm not familiar with the underlying issue here.

File: src/providers/gdal/qgsgdalprovider.cpp [hunk 1]
```diff
@@ -847,19 +847,20 @@ void QgsGdalProvider::readBlock( int bandNo, QgsRectangle  const &extent, int pi
   QgsDebugMsgLevel( QStringLiteral( "tmpXMin = %1 tmpYMax = %2 tmpWidth = %3 tmpHeight = %4" ).arg( tmpXMin ).arg( tmpYMax ).arg( tmpWidth ).arg( tmpHeight ), 5 );
 
   // Allocate temporary block
-  char *tmpBlock = ( char * )qgsMalloc( dataSize * tmpWidth * tmpHeight );
+  size_t bufferSize = dataSize * static_cast<size_t>( tmpWidth ) * static_cast<size_t>( tmpHeight );
```
Comment (rouault) [inline src/providers/gdal/qgsgdalprovider.cpp]: on 32 bit builds, that might overflow, so if we want to be bullet proof, that computation should be done explicitly on quint64 (or whatever QT has for a uint64), cast to size_t, compare the 2 values, and error out if they are different

**Comment by wonder-sk** [comment]
@elpaso regarding this piece of code
https://github.com/qgis/QGIS/blob/master/src/providers/gdal/qgsgdalproviderbase.cpp#L282

I introduced the use of feedback but then disabled it again as explained in the comment below:
```
    // Currently the cancelation is disabled... When RasterIO call is canceled,
    // GDAL returns CE_Failure with error code = 0 (CPLE_None), however one would
    // expect to get CPLE_UserInterrupt to clearly identify that the failure was
    // caused by the cancelation and not that something dodgy is going on.
    // Are both error codes acceptable?
```
I did not go further to investigate whether the above mentioned behavior was correct or whether it was a bug in GDAL - maybe @rouault could shed some more light on how the the return codes from RasterIO are meant to work in case of cancellation.

However at the time when I played with the cancellation I have realized that only very few GDAL drivers actually supported cancellation (e.g. wms for which we have native provider in QGIS) so it became a low priority for me as it would not allow cancellation of common formats like geotiff etc. Maybe that has changed since then...

**Comment by elpaso** [comment]
@wonder-sk thanks for the context, I actually did read the comment but I was working on a different issue and I could not open a new "mental" branch to dig into another one (and I still would prefer not to do that). 

I think that cancellation should be available for this use case: reading a huge raster can take a lot of time and it currently completely freezes the GUI.

As an alternative, we could run the raster calculator in a thread, but that opens new potential issues, and there might still be other situations in which a Raster IO process should be user-interruptable anyway.

So, summarizing, was it only a problem of returned code values? 
It doesn't look like a serious blocker to reintroduce this feature.

File: src/providers/gdal/qgsgdalprovider.cpp (hunk 1 above)
Comment (elpaso) [reply src/providers/gdal/qgsgdalprovider.cpp]: @rouault I'm not sure I got that right, can you please have a look to https://github.com/qgis/QGIS/pull/8550/commits/b3ed482bc8f2ed1408d9ae1e834d53fff771edc9#diff-e743a574c60fb6098a9e6f631e9aadabR851 ?

**Pull request merged**